src/mongo/client/dbclientcursorshimcursorid.cpp

```diff
@@ -36,7 +36,9 @@ namespace mongo  {
             cursor.ns = ele["ns"].String();
 
             if (!ele["firstBatch"].eoo()) {
-                iter = BSONObjIterator(ele["firstBatch"].Obj());
+                BSONObj firstBatch = ele["firstBatch"].Obj();
+                objects_left = firstBatch.nFields();
```

Comment:
ok. scanning 4MB of memory for elements is going to take a while, but if you have >100k collections, well, some things are going to be slower.